specs/experimental/op-contracts-manager.md

```diff
@@ -18,148 +18,262 @@ of governance approved [contract releases] can be found on the
-for future versions of OP Contracts Manager that handle upgrades.
+  [ethereum-lists/chains] repository. This is not enforced onchain, but may matter
+  for future versions of OP Contracts Manager that handle upgrades.
 
 On success, the following event is emitted:
 
 ```solidity
-event Deployed(uint256 indexed l2ChainId, SystemConfig indexed systemConfig);
+event Deployed(uint256 indexed outputVersion, uint256 indexed l2ChainId, address indexed deployer, bytes deployOutput);
 ```
 
 This method reverts on failure. This occurs when:
 
 - The input `l2ChainId` does not comply with the restrictions above.
 - The resulting configuration is not compliant with the [standard configuration].
 
-### Getter Methods
+### Implementation
 
-The following interface defines the available getter methods:
+#### Batch Inbox Address
+
+The chain's [Batch Inbox] address is computed at deploy time using the recommend approach defined
+in the [standard configuration]. This improves UX by removing an input, and ensures uniqueness of
+the batch inbox addresses.
+
+#### Contract Deployments
+
+All contracts deployed by the OP Contracts Manager are deployed with CREATE2.
+
+For singletons the following salt is used:
+
+```solidity
+keccak256(abi.encode(l2ChainId, saltMixer))
+```
+
+For `Proxy` contracts, the following salt is used:
 
 ```solidity
-/// @notice The logic address and initializer selector for an implementation contract.
-struct Implementation {
-  address logic; // Address containing the deployed logic contract.
-  bytes4 initializer; // Function selector for the initializer.
+keccak256(bytes.concat(bytes32(uint256(l2ChainId)), saltMixer, contractName))
+```
+
+The `saltMixer` value is provided as a field in the `DeployInput` struct.
+
+This provides the following benefits:
+
+- Contract addresses for a chain can be derived as a function of chain ID without any RPC calls.
+- Chain ID uniqueness is enforced for free, as a deploy using the same chain ID
+  will result in attempting to deploy to the same address, which is prohibited by
+  the EVM.
+  - This property is contingent on the proxy and `AddressManager` code not
+    changing when OP Contracts Manager is upgraded. Both of these are not planned to
+    change.
+  - The OP Contracts Manager is not responsible for enforcing chain ID uniqueness, so it is acceptable
+    if this property is not preserved in future versions of the OP Contracts Manager.
+
+## Upgrading
+
+### Interface
+
+#### `upgrade`
+
+The `upgrade` method is used by the Upgrade Controller to upgrade the full set of L1 contracts for
+all chains that it controls.
+
+It has the following interface:
+
+```solidity
+/// @notice The full set of inputs to deploy a new OP Stack chain.
+struct DeployInput {
+    Roles roles;
+    uint32 basefeeScalar;
+    uint32 blobBasefeeScalar;
+    uint256 l2ChainId;
+    // The correct type is AnchorStateRegistry.StartingAnchorRoot[] memory,
+    // but OP Deployer does not yet support structs.
```

Comment:
support arrays? `Roles` is a struct

Reply:
The limitation is that neither the [setters](https://github.com/ethereum-optimism/optimism/blob/upgrade-against-registry/packages/contracts-bedrock/scripts/deploy/DeployOPChain.s.sol#L111), nor [getters](https://github.com/ethereum-optimism/optimism/blob/upgrade-against-registry/packages/contracts-bedrock/scripts/deploy/DeployOPChain.s.sol#L162) on the Input/Output contracts that `op-deployer` uses be a complex/reference type, because those I/O contracts are simulated as predeploys in go code, and that simulation does not have the ability to handle the more complicated encodings.

That however is an implementation detail specific to the monorepo, which would not be required for another implementer to know, so I've removed these comments which had been copied from the impl source.

I've also removed the monorepo specific type wrappers from the `DeployInput` and replaced them with their simple type definitions (ie. `Duration` -> `uint64`, `Claim` -> `bytes32`).